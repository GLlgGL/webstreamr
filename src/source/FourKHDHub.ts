import bytes from 'bytes';
import * as cheerio from 'cheerio';
import { BasicAcceptedElems, CheerioAPI } from 'cheerio';
import { AnyNode } from 'domhandler';
import rot13Cipher from 'rot13-cipher';
import { ContentType } from 'stremio-addon-sdk';
import { Context, CountryCode, Meta } from '../types';
import { Fetcher, findCountryCodes, getTmdbId, getTmdbNameAndYear, Id, TmdbId } from '../utils';
import { Source, SourceResult } from './Source';

export class FourKHDHub extends Source {
  public readonly id = '4khdhub';

  public readonly label = '4KHDHub';

  public readonly contentTypes: ContentType[] = ['movie', 'series'];

  public readonly countryCodes: CountryCode[] = [CountryCode.multi];

  public readonly baseUrl = 'https://4khdhub.fans';

  private readonly fetcher: Fetcher;

  public constructor(fetcher: Fetcher) {
    super();

    this.fetcher = fetcher;
  }

  public async handleInternal(ctx: Context, _type: string, id: Id): Promise<SourceResult[]> {
    const tmdbId = await getTmdbId(ctx, this.fetcher, id);

    const pageUrl = await this.fetchPageUrl(ctx, tmdbId);
    if (!pageUrl) {
      return [];
    }

    const html = await this.fetcher.text(ctx, pageUrl);
    const $ = cheerio.load(html);

    if (tmdbId.season) {
      return Promise.all(
        $(`.episode-item`)
          .filter((_i, el) => $('.episode-title', el).text().includes(`S${String(tmdbId.season).padStart(2, '0')}`))
          .map(async (_i, el) => {
            const downloadItemEl = $('.episode-download-item', el)
              .filter((_i, el) => $(el).text().includes(`Episode-${String(tmdbId.episode).padStart(2, '0')}`))
              .get(0);

            return await this.extractSourceResults(ctx, $, downloadItemEl as BasicAcceptedElems<AnyNode>, findCountryCodes($(el).html() as string));
          }).toArray(),
      );
    }

    return Promise.all(
      $(`.download-item`)
        .map(async (_i, el) => await this.extractSourceResults(ctx, $, el, findCountryCodes($(el).html() as string)))
        .toArray(),
    );
  };

  private readonly fetchPageUrl = async (ctx: Context, tmdbId: TmdbId): Promise<URL | undefined> => {
    const [name, year] = await getTmdbNameAndYear(ctx, this.fetcher, tmdbId);

    const searchUrl = new URL(`/?s=${encodeURIComponent(`${name} ${year}`)}`, this.baseUrl);
    const html = await this.fetcher.text(ctx, searchUrl);

    const $ = cheerio.load(html);

    const yearMatches = $(`.movie-card`)
      .slice(0, 2)
      .filter((_i, el) => parseInt($('.movie-card-meta', el).text()) === year);
    const exactTitleMatches = yearMatches
      .filter((_i, el) => $('.movie-card-title', el).text().includes(name));

    return (exactTitleMatches.length ? exactTitleMatches : yearMatches)
      .map((_i, el) => new URL($(el).attr('href') as string, this.baseUrl))
      .get(0);
  };

  private readonly extractSourceResults = async (ctx: Context, $: CheerioAPI, el: BasicAcceptedElems<AnyNode>, countryCodes: CountryCode[]): Promise<SourceResult> => {
    const localHtml = $(el).html() as string;

    const sizeMatch = localHtml.match(/([\d.]+ ?[GM]B)/) as string[];
    const heightMatch = localHtml.match(/\d{3,}p/) as string[];

    const meta: Meta = {
      countryCodes: [...new Set([...countryCodes, ...findCountryCodes(localHtml)])],
      bytes: bytes.parse(sizeMatch[1] as string) as number,
      height: parseInt(heightMatch[0] as string),
      title: $('.file-title, .episode-file-title', el).text().trim(),
    };

    const redirectUrlHubCloud = $('a', el)
      .filter((_i, el) => $(el).text().includes('HubCloud'))
      .map((_i, el) => new URL($(el).attr('href') as string))
      .get(0);

    if (redirectUrlHubCloud) {
      return { url: await this.resolveRedirectUrl(ctx, redirectUrlHubCloud), meta };
    }

    const redirectUrlHubDrive = $('a', el)
      .filter((_i, el) => $(el).text().includes('HubDrive'))
      .map((_i, el) => new URL($(el).attr('href') as string))
      .get(0) as URL;

    const hubDriveHtml = await this.fetcher.text(ctx, await this.resolveRedirectUrl(ctx, redirectUrlHubDrive));
    const $2 = cheerio.load(hubDriveHtml);

    return { url: new URL($2('a:contains("HubCloud")').attr('href') as string), meta };
  };

  private async resolveRedirectUrl(ctx: Context, redirectUrl: URL): Promise<URL> {
    const redirectHtml = await this.fetcher.text(ctx, redirectUrl);
    const redirectDataMatch = redirectHtml.match(/'o','(.*?)'/) as string[];
    const redirectData = JSON.parse(atob(rot13Cipher(atob(atob(redirectDataMatch[1] as string))))) as { o: string };

    return new URL(atob(redirectData['o']));
  }
}
