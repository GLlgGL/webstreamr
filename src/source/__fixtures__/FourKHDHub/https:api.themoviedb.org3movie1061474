{"adult":false,"backdrop_path":"/eU7IfdWq8KQy0oNd4kKXS0QUR08.jpg","belongs_to_collection":{"id":1540907,"name":"Superman (DCU) Collection","poster_path":"/2L0dVc82ZWUgo6Fe2lnJ5qqUDId.jpg","backdrop_path":"/6xQWbcYRHoVDRkyspTy64x8l22m.jpg"},"budget":225000000,"genres":[{"id":878,"name":"Science Fiction"},{"id":12,"name":"Adventure"},{"id":28,"name":"Action"}],"homepage":"https://www.superman.com","id":1061474,"imdb_id":"tt5950044","origin_country":["US"],"original_language":"en","original_title":"Superman","overview":"Superman, a journalist in Metropolis, embarks on a journey to reconcile his Kryptonian heritage with his human upbringing as Clark Kent.","popularity":361.1663,"poster_path":"/ombsmhYUqR4qqOLOxAyr5V8hbyv.jpg","production_companies":[{"id":184898,"logo_path":"/2Z2hiM1ERqFOWRxNxWoJZ8lTxhb.png","name":"DC Studios","origin_country":"US"},{"id":94218,"logo_path":"/kxuhS8wyKtVprOn5nckdHyNuSg1.png","name":"Troll Court Entertainment","origin_country":"US"},{"id":11565,"logo_path":"/odU3l6csuBmXUrzFm6araUgEUHQ.png","name":"The Safran Company","origin_country":"US"},{"id":216687,"logo_path":"/kKVYqekveOvLK1IgqdJojLjQvtu.png","name":"Domain Entertainment","origin_country":"US"}],"production_countries":[{"iso_3166_1":"US","name":"United States of America"}],"release_date":"2025-07-09","revenue":612139671,"runtime":130,"spoken_languages":[{"english_name":"English","iso_639_1":"en","name":"English"}],"status":"Released","tagline":"Look up.","title":"Superman","video":false,"vote_average":7.5,"vote_count":2896}