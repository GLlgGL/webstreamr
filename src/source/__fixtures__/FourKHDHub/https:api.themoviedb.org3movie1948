{"adult":false,"backdrop_path":"/6EX31EmX6lo0SPZB93Ls2ReBfbd.jpg","belongs_to_collection":{"id":64751,"name":"Crank Collection","poster_path":"/wmUzRqcAi0KOW7sqnaAvEQtM8l.jpg","backdrop_path":"/g49FylpoLi9aUPWOqvPCIEezCHN.jpg"},"budget":12000000,"genres":[{"id":28,"name":"Action"},{"id":53,"name":"Thriller"},{"id":80,"name":"Crime"}],"homepage":"","id":1948,"imdb_id":"tt0479884","origin_country":["US"],"original_language":"en","original_title":"Crank","overview":"Chev Chelios, a hit man wanting to go straight, lets his latest target slip away. Then he awakes the next morning to a phone call that informs him he has been poisoned and has only an hour to live unless he keeps adrenaline coursing through his body while he searches for an antidote.","popularity":6.3906,"poster_path":"/rsKmhnvzJezjwC1Ud2Hh37oNpdQ.jpg","production_companies":[{"id":126,"logo_path":"/kADPanRDkKTxncPXXVahvHPUvCj.png","name":"Lakeshore Entertainment","origin_country":"US"},{"id":89394,"logo_path":"/m1Hku3PhgdsQiPO8uVe7szyrxAb.png","name":"RadicalMedia","origin_country":"US"},{"id":2150,"logo_path":null,"name":"GreeneStreet Films","origin_country":"US"},{"id":1632,"logo_path":"/cisLn1YAUuptXVBa0xjq7ST9cH0.png","name":"Lionsgate","origin_country":"US"}],"production_countries":[{"iso_3166_1":"US","name":"United States of America"}],"release_date":"2006-08-31","revenue":42900000,"runtime":88,"spoken_languages":[{"english_name":"English","iso_639_1":"en","name":"English"},{"english_name":"Spanish","iso_639_1":"es","name":"Español"},{"english_name":"Korean","iso_639_1":"ko","name":"한국어/조선말"}],"status":"Released","tagline":"There are a thousand ways to raise your adrenaline. Today, Chev Chelios will need every single one.","title":"Crank","video":false,"vote_average":6.65,"vote_count":3978}