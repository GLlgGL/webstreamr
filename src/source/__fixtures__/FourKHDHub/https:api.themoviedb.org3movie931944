{"adult":false,"backdrop_path":"/j4RUnJz8QCuWkmB8CPykD5UvgBb.jpg","belongs_to_collection":null,"budget":0,"genres":[{"id":18,"name":"Drama"},{"id":9648,"name":"Mystery"},{"id":36,"name":"History"},{"id":27,"name":"Horror"}],"homepage":"","id":931944,"imdb_id":"tt29141112","origin_country":["AT","DE"],"original_language":"de","original_title":"Des Teufels Bad","overview":"In 1750 Austria, a deeply religious woman named Agnes has just married her beloved, but her mind and heart soon grow heavy as her life becomes a long list of chores and expectations. Day after day, she is increasingly trapped in a murky and lonely path leading to evil thoughts, until the possibility of committing a shocking act of violence seems like the only way out of her inner prison.","popularity":2.163,"poster_path":"/ycoXsJomPmPjtCfNweM0UWiTkPY.jpg","production_companies":[{"id":25523,"logo_path":"/hVnHNJUHSxaTpQdWPzC8B5CRvns.png","name":"Ulrich Seidl Filmproduktion","origin_country":"AT"},{"id":23406,"logo_path":"/BTYR4mdM8dQxK0cK0gPVhkXd8j.png","name":"Heimatfilm","origin_country":"DE"},{"id":122,"logo_path":"/pbRxJ8oia1MypvLbukeamObtYr2.png","name":"Coop99 Filmproduktion","origin_country":"AT"}],"production_countries":[{"iso_3166_1":"AT","name":"Austria"},{"iso_3166_1":"DE","name":"Germany"}],"release_date":"2024-03-08","revenue":0,"runtime":121,"spoken_languages":[{"english_name":"German","iso_639_1":"de","name":"Deutsch"}],"status":"Released","tagline":"","title":"The Devil's Bath","video":false,"vote_average":6.603,"vote_count":131}