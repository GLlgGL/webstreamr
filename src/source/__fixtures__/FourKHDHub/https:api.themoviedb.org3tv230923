{"adult":false,"backdrop_path":"/vUVPHEo4ayCO4kkNV4k0PbWPmZS.jpg","created_by":[{"id":4220681,"credit_id":"64de76d9371097011c53058f","name":"Kim Bbang","original_name":"Kim Bbang","gender":0,"profile_path":null}],"episode_run_time":[70],"first_air_date":"2024-04-08","genres":[{"id":18,"name":"Drama"},{"id":10765,"name":"Sci-Fi & Fantasy"},{"id":35,"name":"Comedy"},{"id":9648,"name":"Mystery"}],"homepage":"https://tvn.cjenm.com/ko/Lovely-Runner/","id":230923,"in_production":false,"languages":["ko"],"last_air_date":"2024-05-28","last_episode_to_air":{"id":5148777,"name":"I Remember Everything","overview":"Sun-jae's memory comes back and now he remembers everything that happened between Sol and him. Thanks to Tae-sung, the ill-fated relations are resolved and the couple can be together. Sun-jae wants to propose to Sol, but finding the right time is difficult. Sun-ae decides to attend Sol’s nephew’s first birthday party and meet Sol’s family.","vote_average":0.0,"vote_count":1,"air_date":"2024-05-28","episode_number":16,"episode_type":"finale","production_code":"","runtime":77,"season_number":1,"show_id":230923,"still_path":"/hcFyYk5LGmkqTSbxRnQqej3mszO.jpg"},"name":"Lovely Runner","next_episode_to_air":null,"networks":[{"id":866,"logo_path":"/aRDq8zBrX3YLpHSfueNQBkNnn7b.png","name":"tvN","origin_country":"KR"}],"number_of_episodes":16,"number_of_seasons":1,"origin_country":["KR"],"original_language":"ko","original_name":"선재 업고 튀어","overview":"Right after Ryu Sun-jae, a top star, ends his life, Im Sol, his top fan, somehow ends up at a time when they were in high school and tries to protect him. A fantasy romance unfolds where people who missed each other in time finally meet.","popularity":10.0734,"poster_path":"/xJQyrif5M4UMoVBrBlwUabtaRxB.jpg","production_companies":[{"id":36225,"logo_path":"/haNfBy6ZS1a1teY9JXNHHkFhZEj.png","name":"Bon Factory","origin_country":"KR"},{"id":199265,"logo_path":"/mv2yDqAgT1ea1eTXZMDr89zX95M.png","name":"CJ ENM Studios","origin_country":"KR"}],"production_countries":[{"iso_3166_1":"KR","name":"South Korea"}],"seasons":[{"air_date":"2024-04-08","episode_count":16,"id":349357,"name":"Season 1","overview":"Im Sol is devastated by the sudden and tragic death of her favorite star, Ryu Sun-jae, a former swimmer turned K-pop idol. But in the midst of her anguish, she magically time slips back to 2008 – when Sun-jae is still only 19 years old! Delighted that she has a chance to help him avert his unhappy future fate, Sol makes it her mission to change the future... and maybe find love along the way.","poster_path":"/xJQyrif5M4UMoVBrBlwUabtaRxB.jpg","season_number":1,"vote_average":10.0}],"spoken_languages":[{"english_name":"Korean","iso_639_1":"ko","name":"한국어/조선말"}],"status":"Ended","tagline":"I will protect both you and me.","type":"Scripted","vote_average":8.578,"vote_count":210}