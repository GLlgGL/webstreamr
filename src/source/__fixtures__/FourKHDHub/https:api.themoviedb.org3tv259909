{"adult":false,"backdrop_path":"/8J0SjKToFBbg8dKpocVOCsrC77C.jpg","created_by":[{"id":1212409,"credit_id":"67bdd41976904c59c85e9f6c","name":"Clyde Phillips","original_name":"Clyde Phillips","gender":2,"profile_path":"/ue6gnd9FRbl5wKZ9yHVtWp6ra4y.jpg"}],"episode_run_time":[53],"first_air_date":"2025-07-13","genres":[{"id":80,"name":"Crime"},{"id":18,"name":"Drama"},{"id":9648,"name":"Mystery"}],"homepage":"https://www.paramountplus.com/shows/dexter-resurrection/","id":259909,"in_production":true,"languages":["en"],"last_air_date":"2025-08-31","last_episode_to_air":{"id":6201599,"name":"Touched by an Ángel","overview":"Prater and Charley's revelation forces Dexter to go to extreme lengths.","vote_average":8.8,"vote_count":8,"air_date":"2025-08-31","episode_number":9,"episode_type":"standard","production_code":"","runtime":48,"season_number":1,"show_id":259909,"still_path":"/35X0k5JHwJtehW37ELaVWOHGbb0.jpg"},"name":"Dexter: Resurrection","next_episode_to_air":{"id":6201600,"name":"And Justice for All…","overview":"Charley's allegiance to Prater is pushed to its limit, forcing him to make a choice. Dexter and Harrison find themselves in a final confrontation, facing the consequences of their darkness. Season finale.","vote_average":0.0,"vote_count":0,"air_date":"2025-09-07","episode_number":10,"episode_type":"finale","production_code":"","runtime":null,"season_number":1,"show_id":259909,"still_path":"/WTBlw7Sq76wFABHwMbnxZJuZTn.jpg"},"networks":[{"id":6631,"logo_path":"/zFEsDBjBEj5OiM0FDRYY1NnG7a9.png","name":"Paramount+ with Showtime","origin_country":"US"}],"number_of_episodes":10,"number_of_seasons":1,"origin_country":["US"],"original_language":"en","original_name":"Dexter: Resurrection","overview":"Dexter Morgan awakens from a coma to find Harrison gone without a trace. Realizing the weight of what he put his son through, Dexter sets out for New York City, determined to find him and make things right. But closure won't come easy. When Miami Metro's Angel Batista arrives with questions, Dexter realizes his past is catching up to him fast. As father and son navigate their own darkness in the city that never sleeps, they soon find themselves deeper than they ever imagined - and that the only way out is together.","popularity":80.1526,"poster_path":"/hIawSocuwqgNeRf3JuKuxgMHmSC.jpg","production_companies":[{"id":17595,"logo_path":null,"name":"Clyde Phillips Productions","origin_country":"US"},{"id":247697,"logo_path":null,"name":"Sal Centric","origin_country":""},{"id":238896,"logo_path":null,"name":"Counterpart Studios","origin_country":"US"},{"id":238897,"logo_path":null,"name":"Showtime Studios","origin_country":"US"}],"production_countries":[{"iso_3166_1":"US","name":"United States of America"}],"seasons":[{"air_date":"2025-07-13","episode_count":10,"id":403778,"name":"Season 1","overview":"Set a few weeks after Dexter: New Blood, Dexter Morgan chases a missing Harrison to New York City, where Captain Angel Batista is hot on their trail from way down in Miami.","poster_path":"/hRD9kvYLbySRvtt9Nh7QK6XECZX.jpg","season_number":1,"vote_average":8.8}],"spoken_languages":[{"english_name":"English","iso_639_1":"en","name":"English"}],"status":"Returning Series","tagline":"He's alive & killing it.","type":"Scripted","vote_average":8.9,"vote_count":179}