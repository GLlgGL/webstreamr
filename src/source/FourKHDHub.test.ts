import { createTestContext } from '../test';
import { FetcherMock, TmdbId } from '../utils';
import { FourKHDHub } from './FourKHDHub';

const ctx = createTestContext();

describe('FourKHDHub', () => {
  let source: FourKHDHub;

  beforeEach(() => {
    source = new FourKHDHub(new FetcherMock(`${__dirname}/__fixtures__/FourKHDHub`));
  });

  test('handle non-existent devil\'s bath 2024 gracefully', async () => {
    const streams = await source.handle(ctx, 'movie', new TmdbId(931944, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });

  test('handle superman 2025', async () => {
    const streams = await source.handle(ctx, 'movie', new TmdbId(1061474, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });

  test('handle dark 2017 s01e02', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(70523, 1, 2));
    expect(streams).toMatchSnapshot();
  });

  test('handle dexter resurrection 2025 s01e01', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(259909, 1, 1));
    expect(streams).toMatchSnapshot();
  });

  test('handle dexter original sin 2024 s01e01', async () => {
    const streams = await source.handle(ctx, 'series', new TmdbId(219937, 1, 1));
    expect(streams).toMatchSnapshot();
  });

  test('handle crayon shin-chan 1993', async () => {
    const streams = await source.handle(ctx, 'movie', new TmdbId(128868, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });

  test('handle crayon shin-chan 1998', async () => {
    const streams = await source.handle(ctx, 'movie', new TmdbId(128875, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });

  test('handle crank 2006', async () => {
    const streams = await source.handle(ctx, 'movie', new TmdbId(1948, undefined, undefined));
    expect(streams).toMatchSnapshot();
  });
});
