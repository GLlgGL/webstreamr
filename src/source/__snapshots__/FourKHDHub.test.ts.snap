// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FourKHDHub handle crank 2006 1`] = `
[
  {
    "meta": {
      "bytes": 17255031111,
      "countryCodes": [
        "en",
        "hi",
      ],
      "height": 1080,
      "title": "Crank (2006) Director's Cut 1080p BluRay REMUX VC-1 [Hindi AMZN DDP 5.1 ~ 640kbps + English DTS-HD HRA 5.1] ESubs ~ (Arrow) FraMeSToR.mkv",
    },
    "url": "https://hubcloud.one/drive/lqa8qw881tasohi",
  },
  {
    "meta": {
      "bytes": 10479720202,
      "countryCodes": [
        "en",
        "hi",
      ],
      "height": 1080,
      "title": "Crank (2006) Director's Cut 1080p BluRay x264 [Hindi AMZN DDP 5.1 ~ 640kbps + English DTS] ESubs ~ (Arrow) CtrlHD.mkv",
    },
    "url": "https://hubcloud.one/drive/eijzmr9jiwy1r4z",
  },
  {
    "meta": {
      "bytes": 4176855695,
      "countryCodes": [
        "en",
        "hi",
      ],
      "height": 1080,
      "title": "Crank (2006) Director's Cut 1080p 10bit BluRay x265 HEVC [Hindi AMZN DDP 5.1 ~ 640Kbps + English DD 5.1] ESubs ~ (Arrow) Chivaman.mkv",
    },
    "url": "https://hubcloud.one/drive/sc0yyca6hq11y0r",
  },
]
`;

exports[`FourKHDHub handle crayon shin-chan 1993 1`] = `
[
  {
    "meta": {
      "bytes": 28550795100,
      "countryCodes": [
        "hi",
        "ja",
      ],
      "height": 1080,
      "title": "Crayon Shinchan Action Kamen vs Demon (1993) 1080p BluRay REMUX AVC [Hindi DD 2.0 + Japanese FLAC 2.0] (PitiPati-Ionicboy) (4kHDHub.com).mkv",
    },
    "url": "https://hubcloud.one/drive/bffzqlpqfllfcld",
  },
  {
    "meta": {
      "bytes": 10383083438,
      "countryCodes": [
        "hi",
        "ja",
      ],
      "height": 1080,
      "title": "Crayon Shinchan Action Kamen vs Demon (1993) 1080p BluRay [Hindi DD 2.0 + Japanese DD 2.0] x264 (WiKi-Ionicboy) (4kHDHub.com).mkv",
    },
    "url": "https://hubcloud.one/drive/zsn5msss5usmasw",
  },
  {
    "meta": {
      "bytes": 5776731013,
      "countryCodes": [
        "hi",
        "ja",
      ],
      "height": 1080,
      "title": "Crayon Shinchan Action Kamen vs Demon (1993) 1080p 10bit BluRay HEVC [Hindi DD 2.0 + Japanese DD 2.0] H.265 (WiKi-Ionicboy) (4kHDHub.com).mkv",
    },
    "url": "https://hubcloud.one/drive/vlv5sssjtztvrsc",
  },
]
`;

exports[`FourKHDHub handle crayon shin-chan 1998 1`] = `[]`;

exports[`FourKHDHub handle dark 2017 s01e02 1`] = `
[
  {
    "meta": {
      "bytes": 2126008811,
      "countryCodes": [
        "de",
        "en",
      ],
      "height": 1080,
      "title": "Dark.S01E02.Lies.1080p.NF.WEB-DL.English.DDP5.1-German.DDP5.1.Atmos.H.264-4kHdHub.Com.mkv",
    },
    "url": "https://hubcloud.one/drive/rq7tg4llqgks1tq",
  },
  {
    "meta": {
      "bytes": 4348654387,
      "countryCodes": [
        "de",
        "en",
      ],
      "height": 2160,
      "title": "Dark.S01E02.Lies.2160p.NF.WEB-DL.DUAL.DDP5.1.Atmos.H.265-4kHdHub.Com.mkv",
    },
    "url": "https://hubcloud.one/drive/p1uulgcc4g4k9cq",
  },
]
`;

exports[`FourKHDHub handle dexter original sin 2024 s01e01 1`] = `
[
  {
    "meta": {
      "bytes": 5733781340,
      "countryCodes": [
        "en",
      ],
      "height": 2160,
      "title": "Dexter.Original.Sin.S01E01.2160p.SKST.WEB-DL.H.265.DDP5.1-Tyrell.mkv",
    },
    "url": "https://hubcloud.one/drive/idt1evqfuviqiei",
  },
  {
    "meta": {
      "bytes": 5143223336,
      "countryCodes": [
        "en",
      ],
      "height": 2160,
      "title": "Dexter.Original.Sin.S01E01.2160p.PMTP.WEB-DL.DDP5.1.DV.H.265-Tyrell.mkv",
    },
    "url": "https://hubcloud.one/drive/v6yd0pgnyteceev",
  },
  {
    "meta": {
      "bytes": 5046586572,
      "countryCodes": [
        "en",
      ],
      "height": 2160,
      "title": "Dexter.Original.Sin.S01E01.2160p.PMTP.WEB-DL.DDP5.1.HDR.H.265-Tyrell.mkv",
    },
    "url": "https://hubcloud.one/drive/6f11sgo1qosl60k",
  },
  {
    "meta": {
      "bytes": 5057323991,
      "countryCodes": [
        "en",
      ],
      "height": 2160,
      "title": "Dexter.Original.Sin.S01E01.2160p.PMTP.WEB-DL.DDP5.1.DV.HDR.H.265-Tyrell.mkv",
    },
    "url": "https://hubcloud.one/drive/ugrbggibuhb1ggd",
  },
]
`;

exports[`FourKHDHub handle dexter resurrection 2025 s01e01 1`] = `
[
  {
    "meta": {
      "bytes": 6012954214,
      "countryCodes": [
        "en",
      ],
      "height": 2160,
      "title": "Dexter.Resurrection.S01E01.A.Beating.Heart.2160p.AMZN.WEB-DL.English.DDP5.1.H.265-4kHdHub.Com.mkv",
    },
    "url": "https://hubcloud.one/drive/sa0mjerabpraxas",
  },
  {
    "meta": {
      "bytes": 3285649981,
      "countryCodes": [
        "en",
      ],
      "height": 1080,
      "title": "Dexter.Resurrection.S01E01.A.Beating.Heart.1080p.AMZN.WEB-DL.English.DDP5.1.H.264-4kHdHub.Com.mkv",
    },
    "url": "https://hubcloud.one/drive/xzdxd7iexxxxx2v",
  },
  {
    "meta": {
      "bytes": 935109591,
      "countryCodes": [
        "en",
      ],
      "height": 1080,
      "title": "Dexter.Resurrection.S01E01.A.Beating.Heart.1080p.AMZN.WEB-DL.English.DDP5.1.H.265-4kHdHub.Com.mkv",
    },
    "url": "https://hubcloud.one/drive/2hs2gt33gcggocc",
  },
]
`;

exports[`FourKHDHub handle lovely runner 2024 s01e01 1`] = `
[
  {
    "meta": {
      "bytes": 3586297692,
      "countryCodes": [
        "hi",
        "ko",
      ],
      "height": 1080,
      "title": "Lovely.Runner.S01E01.1080p.AMZN.WEB-DL.Hindi.DDP2.0-Korean.DDP2.0.H.264-4kHdHub.Com.mkv",
    },
    "url": "https://hubcloud.one/drive/8t7i1lrytgz56eo",
  },
  {
    "meta": {
      "bytes": 10340133765,
      "countryCodes": [
        "hi",
        "ko",
      ],
      "height": 1080,
      "title": "Lovely Runner S01E01 1080p WEB-DL [Hindi DDP 2.0 + Korean DDP 5.1] x264 (PandaMoon-LUMiX).mkv",
    },
    "url": "https://hubcloud.one/drive/171fxx1xmbezm1r",
  },
]
`;

exports[`FourKHDHub handle non-existent devil's bath 2024 gracefully 1`] = `[]`;

exports[`FourKHDHub handle superman 2025 1`] = `
[
  {
    "meta": {
      "bytes": 65326452572,
      "countryCodes": [
        "en",
        "hi",
      ],
      "height": 2160,
      "title": "Superman (2025) 2160p UHD BluRay REMUX DoVi HDR HEVC [Hin-Tam-Tel DDP 5.1 + English TrueHD.Atmos 7.1] (CiNEPHiLES-Ionicboy).mkv",
    },
    "url": "https://hubcloud.one/drive/pxpz19pzzzgi99y",
  },
  {
    "meta": {
      "bytes": 26940182364,
      "countryCodes": [
        "en",
        "hi",
      ],
      "height": 1080,
      "title": "Superman (2025) 1080p BluRay REMUX AVC [Hin-Tam-Tel DDP 5.1 + English TrueHD.Atmos 7.1] (FraMeSToR-Ionicboy).mkv",
    },
    "url": "https://hubcloud.one/drive/2nx1rzuzhtokphj",
  },
  {
    "meta": {
      "bytes": 26671746908,
      "countryCodes": [
        "en",
        "hi",
      ],
      "height": 2160,
      "title": "Superman (2025) IMAX 2160p MA WEB-DL DV HDR10+ 10bit HEVC [Hindi-Tamil-Telugu DDP 5.1 + English DDP Atmos 5.1] x265 (126811-LUMiX).mkv",
    },
    "url": "https://hubcloud.one/drive/vhcdcl3dyw0g00f",
  },
  {
    "meta": {
      "bytes": 26661009489,
      "countryCodes": [
        "en",
        "hi",
      ],
      "height": 2160,
      "title": "Superman (2025) IMAX 2160p MA WEB-DL SDR 10bit HEVC [Hindi-Tamil-Telugu DDP 5.1 + English DDP Atmos 5.1] x265 (HONE-LUMiX).mkv",
    },
    "url": "https://hubcloud.one/drive/rwit2wctpiuiw51",
  },
  {
    "meta": {
      "bytes": 23665269800,
      "countryCodes": [
        "en",
        "hi",
      ],
      "height": 1080,
      "title": "Superman (2025) IMAX 1080p MA WEBRip [Hindi-Tamil-Telugu DDP 5.1 + English DDP Atmos 5.1] x264 (HiDt-LUMiX).mkv",
    },
    "url": "https://hubcloud.one/drive/2i1vnx0nnxxsnnz",
  },
  {
    "meta": {
      "bytes": 21775484190,
      "countryCodes": [
        "en",
        "hi",
      ],
      "height": 1080,
      "title": "Superman (2025) 1080p 10bit UHD BluRay DoVi HDR HEVC [Hin-Tam-Tel DDP 5.1 + English DDP 7.1] H.265 (ATELiER-Ionicboy).mkv",
    },
    "url": "https://hubcloud.one/drive/oi7nukpci9igocg",
  },
  {
    "meta": {
      "bytes": 10307921510,
      "countryCodes": [
        "en",
        "hi",
      ],
      "height": 1080,
      "title": "Superman (2025) IMAX 1080p MA WEB-DL [Hindi-Tamil-Telugu DDP 5.1 + English DDP Atmos 5.1] x264 (BYNDR-LUMiX).mkv",
    },
    "url": "https://hubcloud.one/drive/ntjnpq5qa5tqtna",
  },
  {
    "meta": {
      "bytes": 7451768258,
      "countryCodes": [
        "en",
        "hi",
      ],
      "height": 1080,
      "title": "Superman (2025) IMAX 1080p DS4K MA WEBRip 10bit HEVC [Hindi-Tamil-Telugu DDP 5.1 + English DDP Atmos 5.1] x265 (StOiC-LUMiX).mkv",
    },
    "url": "https://hubcloud.one/drive/f8a1y8218xq1xl1",
  },
]
`;
